### ISSUE TITLE
QueryTypes: Defining query on hierarchy of query types (base and derived) doesn't work

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#11789]
repro:

```cs
    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();

                var f1 = new Faction { Name = "Skeliege" };
                var f2 = new Faction { Name = "Monsters" };
                var f3 = new Faction { Name = "Nilfgaard" };
                var f4 = new Faction { Name = "Northern Realms" };
                var f5 = new Faction { Name = "Scioia'tael" };

                var l11 = new Leader { Faction = f1, Name = "Bran Tuirseach" };
                var l12 = new Leader { Faction = f1, Name = "Crach an Craite" };
                var l13 = new Leader { Faction = f1, Name = "Eist Tuirseach" };
                var l14 = new Leader { Faction = f1, Name = "Harald the Cripple" };

                ctx.Factions.AddRange(f1, f2, f3, f4, f5);
                ctx.Leaders.AddRange(l11, l12, l13, l14);
                ctx.SaveChanges();
            }

            using (var ctx = new MyContext())
            {
                ctx.LeadersQuery.ToList();
            }
        }
    }

    public class MyContext : DbContext
    {
        public DbSet<Faction> Factions { get; set; }
        public DbSet<Leader> Leaders { get; set; }

        public DbQuery<LeaderQuery> LeadersQuery { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=.;Database=QueryTypesRepro;Trusted_Connection=True;MultipleActiveResultSets=True");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Query<LeaderQuery>().ToQuery(() => Set<Leader>().Select(l => new LeaderQuery { Name = l.Name, Faction = l.Faction.Leaders.OrderBy(ll => ll.Id).Select(ll => ll.Faction).FirstOrDefault() } ));
            modelBuilder.Query<LeaderQueryBase>().ToQuery(() => Set<Leader>().Select(l => new LeaderQueryBase { Name = l.Name }));
        }
    }

    public class Faction
    {
        public int Id { get; set; }
        public string Name { get; set; }

        public List<Leader> Leaders { get; set; }
    }

    public class FactionQuery
    {
        public string Name { get; set; }
    }

    public class Leader
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public Faction Faction { get; set; }
    }

    public class LeaderQuery : LeaderQueryBase
    {
        public Faction Faction { get; set; }
    }

    public class LeaderQueryBase
    {
        public string Name { get; set; }
    }
```

this throws:

```
The query model's result type cannot be changed to 'Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1[QueryTypesTests.LeaderQuery]'. The result type may only be overridden and set to values compatible with the ResultOperators' current data type ('System.Linq.IQueryable`1[QueryTypesTests.LeaderQueryBase]').

   at Remotion.Linq.QueryModel.GetOutputDataInfo()
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.SingleResultToSequence(QueryModel queryModel, Type type) in D:\git\EntityFrameworkCore\src\EFCore\Query\EntityQueryModelVisitor.cs:line 646
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel) in D:\git\EntityFrameworkCore\src\EFCore\Query\EntityQueryModelVisitor.cs:line 177
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel) in D:\git\EntityFrameworkCore\src\EFCore\Storage\Database.cs:line 70
```

QM:

```
from Leader l in DbSet<Leader>
select new LeaderQueryBase{ Name = [l].Name }
```

We seem to be using only the defining query for LeaderQueryBase, when asked to construct `DbQuery<LeaderQuery>`

**Comment by ajcvickers** [comment]
Triage: since it is not clear exactly what the behaviors here could be, the plan for 2.1 is to block in model validation any attempt to set a a defining query on a QueryType that is not the base.

**Comment by ajcvickers** [comment]
This is approved for RTM, but please wait for the branch to open for RTM before merging.

**Comment by AndriySvyryd** [comment]
Fixed in 6be065ceceac92384846f090c2466fbc2ad952fc
